### ISSUE TITLE
Rotated Bounding Boxes

### ISSUE THREAD

**Issue by oke-aditya** [pytorch/vision#2761]
## 🚀 Feature
A bit unsure about this feature.

Support Rotated Bounding Boxes in Torchvision.

## Motivation

There is recent research on Rotated Bounding boxes which provides better detection results. I am not able to find highly cited results but a few of them are
1. [Rotation Invariant Detector](https://arxiv.org/pdf/1711.09405v1.pdf)
2. [SCRDET](https://openaccess.thecvf.com/content_ICCV_2019/papers/Yang_SCRDet_Towards_More_Robust_Detection_for_Small_Cluttered_and_Rotated_ICCV_2019_paper.pdf) ICCV 2019.

I'm not sure for more papers. I think this is slightly new topic, and needs a bit more research.

## Pitch

1. Support operations for Rotated Bounding boxes, NMS, ROI Align. IoU.
2. Support Augmentation (if needed) for Rotated Bounding boxes.
3. Support such Datasets (if any) that use Rotated Boxes. 

If possible we can also support Rotated models. In my opinion it might not be very feasible. 
Since it will take a lot of time and maintenance for these models.

## Alternatives

Currently, Detectron2 has great support for all the above. These operations are implemented in C++.

## Additional context

Looks tricky and challenging. I guess it might also be too early to think about this. I think it needs a bit more research and some baseline.

**Comment by fmassa** [comment]
Hi,

This is a good thing to consider adding to torchvision, but as you mentioned it might involve quite a few changes so it won't be easy to do.

If you are interested in having support for rotated bounding boxes, please "tumbs up" the main post.

**Comment by zhiqwang** [comment]
Hi @fmassa , @oke-aditya , I have noticed that detectron2 have implemented `nms_rotated` [here](https://github.com/facebookresearch/detectron2/blob/6176a25b017b92dbe833d49987526d6e339d1d8a/detectron2/layers/nms.py#L33) and `ROIAlignRotated` [here](https://github.com/facebookresearch/detectron2/blob/6176a25b017b92dbe833d49987526d6e339d1d8a/detectron2/layers/roi_align_rotated.py#L50). And the auther of detectron2 noted that

> # Note: this function (nms_rotated) might be moved into
> # torchvision/ops/boxes.py in the future

Maybe this is an opportune moment to do this?

cc @ppwwyyxx

**Comment by oke-aditya** [comment]
Let's discuss this @zhiqwang

There are a few points about this what I think.
1. Detectron has a class for [Boxes ](https://github.com/facebookresearch/detectron2/blob/6176a25b017b92dbe833d49987526d6e339d1d8a/detectron2/structures/boxes.py#L133) as well as for [rotated bounding boxes](https://github.com/facebookresearch/detectron2/blob/6176a25b017b92dbe833d49987526d6e339d1d8a/detectron2/structures/rotated_boxes.py#L11).
We don't have any such classes in torchvision. This keeps abstraction to Tensor level and makes these ops very easy to use.

For Rotated boxes, we would need to implement these common operations.

Both torchvision and detectron2 represent bounding boxes as `(x1, y1, x2, y2)` for Non rotated. We can covert them though, but all the operations are implmented for this format only.

Detectron2 represents rotated boxes as `(x_center, y_center, width, height, angle)`.
I think this could be standard for torchvision too.

Some proposals for ops

1. For ops such as `box_area`, `clip_boxes_to_image` we can overload the to accept both `Tensor[N, 4]` for non rotated case and `Tensor[N, 5]` . This would keep them a bit less in number. Else we would have ops like `rotated_box_area`, `rotated .. `. Since we don't represnet boxes as classes. Both are fine though

2. Some ops are done using C++ in Detectron2 like @zhiqwang mentioned. Unsure about them.

3. We need to provide rotated bounding box conversions. This is tricky again we can resort to 1. overload the `Tensor` to `Tensor[N, 4]` for non rotated and `Tensor[N, 5]` for rotated. Or we can provide `rotated_box_convert`.

I think Rotated bounding box is a more generic case of non rotated case. Case where angle is 0 represents normal boxes. (Please correct if I'm wrong).

4. We could provide a simple utility to simply convert a rotated box to non rotated case. (Simply tilt back the angle to 0 and remove that dim)

My thoughts might be naive. I guess this needs better thoughts.

**Comment by zhiqwang** [comment]
Hi @oke-aditya , the points you highlight (#2710) makes me clear of the implementation of boxes conversion in `torchvison`.

> Both `torchvision` and `detectron2` represent bounding boxes as `(x1, y1, x2, y2)` for Non rotated. We can covert them though, but all the operations are implmented for this format only.

I think that one reason both `torchvision` and `detectron2` uses the `(x1, y1, x2, y2)` format to represent the rectangle bounding box is associated to the utilization of COCO datasets. But for rotated bounding boxes datasets, there are not a standard/benchmark datasets now? I have encountered two tasks using rotated bounding boxes. One is the OCR task - ICDAR 2015, and the other one is [aerial image detection](https://captain-whu.github.io/DOTA/). The two uses the polygon format `(x1, y1, x2, y2, x3, y3, x4, y4)` (but they are rotated boxes actually?) So one must converted the polygon format to `(x_center, y_center, width, height, angle)` if they want to use detectron2.

And as @vfdev-5 , @pmeier and @oke-aditya mentioned in #2753 . OCR is a higher level application and thus should be in a separate package which might depend on `torchvision`. But if there are operations like `nms_rotated` and `pairwise_iou_rotated` implemented in `torchvision`. It will be convenient for other users.

**Comment by oke-aditya** [comment]
Let's have more thoughts from @fmassa

**Comment by sugatoray** [comment]
## Displaying bounding boxes for rotated rectangles.

Given the rotated blue rectangular region, how to show the label?

If the labels are rotated as well, it will be quite challenging to rotate your head everytime to read the labels. On the other hand, if the label is shown withing the non-rotated bounding box (green), which also contains the target rotated rectangle (blue), then it is clear where the rotated bounding box is and is also easy to read.

So, each bounding box then is composed of two things:

- outer non-rotated bounding box (green), which contains the label
- inner rotated bounding box (blue), which shows the target ROI.
  - format: 
    a. `[(x1, y1), (x2, y2), (x3, y3), (x4, y4)]`   
    b. `[x_center, y_center, height, width, angle]`

But, any format specification should be adequate to determine both bounding boxes and draw them as well.

![image](https://user-images.githubusercontent.com/10201242/121791493-d77de700-cbaf-11eb-9ef2-a64fa1083c23.png)

**Comment by ashnair1** [comment]
Are there plans to port the rotated ops from detectron2 to torchvision?

**Comment by oke-aditya** [comment]
@pmeier I guess with the new Features API this may be possible in near future?

**Comment by pmeier** [comment]
Yes, in theory we could have `features.RotatedBoundingBox` with shape `(*, 5)` where 4 items represent the coordinates and one represents the angle. For now, we already implemented `rotate_bounding_box`:

https://github.com/pytorch/vision/blob/149edda463b54b3eabe989e260a839727c89d099/torchvision/prototype/transforms/functional/_geometry.py#L573

This actually rotates the coordinates and thus the result is a "proper" (for the lack of a better term) bounding box. Thus, the bounding box may change shape.

Excuse my ignorance here, but is rotation invariance that important? What about all the other transformations that change the shape of bounding boxes, like affine, elastic, perspective, and so on? From my, admittedly limited, perspective it seems weird that rotation is singled out here. Or to put differently, having a `features.RotatedBoundingBox` in the library would be odd, since it behaves just like `features.BoundingBox` except for one transformation.

**Comment by oke-aditya** [comment]
@pmeier 
We might need to look into Detectron2 for some Inspiration. As I know operations such as NMS and IoUs etc differ for Rotated boxes.

**Comment by pmeier** [comment]
Maybe I misunderstood the context here: are we talking about rotated bounding boxes at the transforms or at the model stage?

**Comment by oke-aditya** [comment]
Since I had created the issue, my initial idea was that it isn't exactly to do with transforms  but we would need a few small transforms like say horizontal flip etc that need both image and the box. 

But other than that mostly it's on the ops side.

**Comment by datumbox** [comment]
This is a bigger piece of work that we should consider after H2. Philip is right to say that our roadmap is currently full and this is quite a sizeable piece of work to squeeze in. As Aditya said, this is not just about adding support for transforms but also for operators. I would be in favour of reviewing the support of this as a whole as we would need to provide Transforms, Ops and Models that are compatible. IMO we should avoid adding more partial support for features in TorchVision; we already have quite a few things that are partially supported (Keypoints, Video Decoding etc). We are currently investing on closing any high priority gaps on the already supported tasks. Once this is done, we should stack rank big proposed features to see which of them are worth investing into. The situation I would like to avoid is having some support of Rotated BBoxes in some Transforms/Ops/Models but not on others. This is likely to make our Detection API more complex and harder to revamp.

**Comment by NicolasHug** [comment]
This will be supported in 0.23 that we plan to release this week.

**Comment by oke-aditya** [comment]
This sounds so fantastic!! After 5 years.

**Comment by Nebulised** [comment]
Hi, just to confirm this is just support for rotated bounding boxes with torchvision transforms. It does not include support for ops such as rotated roi align?

**Comment by NicolasHug** [comment]
That's correct @Nebulised , only transforms for now, but ops will come soon!

**Comment by Nebulised** [comment]
> That's correct [@Nebulised](https://github.com/Nebulised) , only transforms for now, but ops will come soon!

No worries, great that the Torchvision team are working on this and are looking forward to the advancements made. At least for now we have the Detectron 2 ops.
